### PR TITLE
fix: ABI decoder crashing on certain inputs

### LINKED ISSUE #1108: TOB-FUEL-32: ABI decoder is crashing on certain inputs
## Description
A fuzzing campaign for the ABI decoder revealed several crashes. This finding serves as an umbrella finding to list all the different crashes we observed. The harness for the fuzzer is shown in the following figure.

#### Figure 32.1: Fuzz harness

```rust
#[test_fuzz::test_fuzz]
pub fn fuzz_abi_decode(bytes: &[u8], param_type: ParamType)
{
    ABIDecoder::decode_single(&param_type, bytes);
}
```

The findings are relevant for use cases where untrusted ABI definitions are parsed. All the findings below are due to malformed type definitions. These types are not controlled by Sway programs, but by the compiler which emits an ABI definition.
This finding is Undetermined because it was not part of the initial scope.

### Division of zero
The following unit test panics with attempt to calculate the remainder with a divisor of zero at fuels-rs/packages/fuels-core/src/types/param_types.rs:76.

#### Figure 32.2: Unit test that panics
```rust
#[test]
pub fn div_zero() {
    use crate::types::param_types::ParamType::*;
    fuzz_abi_decode(&[], Vector(Box::new(Array(Box::new(U16), 0))));
}
```

### Attempt to multiply with overflow
The following test panics at fuels-rs/packages/fuels-core/src/codec/abi_decoder.rs:256 with attempt to multiply with overflow.

#### Figure 32.3: Unit test that panics

```rust
#[test]
pub fn multiplay_overflow_enum() {
    use crate::types::enum_variants::EnumVariants;
    use crate::types::param_types::ParamType::*;
    fuzz_abi_decode(
        &[0, 8, 8, 8, 9, 8, 0, 8, 8, 8, 8, 8, 15, 8, 8, 8],
        Enum {
            variants: EnumVariants {
                param_types: vec![
                    Array(Box::new(Array(Box::new(RawSlice), 8)),
576469587185895432),
                    B256,
                    B256,
                    B256,
                    B256,
                    B256,
                    B256,
                    B256,
                    B256,
                    B256,
                    B256,
      ], },
                  generics: vec![U16],
              },
      ); 
}
```

The next panic is caused by fuels-rs/packages/fuels-core/src/types/param_types.rs:76.

#### Figure 32.4: Unit test that panics

```rust
#[test]
pub fn multiplay_overflow_vector() {
    use crate::types::param_types::ParamType::*;
    fuzz_abi_decode(
        &[8, 8, 10, 7, 229, 8, 8, 8],
        Vector(Box::new(Array(Box::new(Unit), 2308103648053880071))),
    );
}
```

The next unit tests panics at fuels-rs/packages/fuels-core/src/types/param_types.rs:155.

#### Figure 32.5: Unit test that panics

```rust
#[test]
pub fn multiplay_overflow_arith() {
    use crate::types::enum_variants::EnumVariants;
    use crate::types::param_types::ParamType;
    use crate::types::param_types::ParamType::*;
    let mut typ: ParamType = U16;
    for _ in 0..50 {
        typ = Array(Box::new(typ), 8);
    }
        fuzz_abi_decode(
            &[0, 8, 8, 51, 51, 51, 51, 51, 51, 51, 3, 8, 15, 8, 8, 8],
            Enum {
                variants: EnumVariants {
                    param_types: vec![typ],
    },
                generics: vec![U16],
            },
    ); 
}
```

### Capacity overflow
The following unit tests panics with the message capacity overflow at fuels-rs/packages/fuels-core/src/codec/abi_decoder.rs:138.

#### Figure 32.6: Unit test that panics

```rust
#[test]
pub fn capacity_overflow() {
    use crate::types::param_types::ParamType::*;
    fuzz_abi_decode(
        &[13, 0, 1, 0, 0, 106, 242, 8],
        Array(
            Box::new(Array(Box::new(Tuple(vec![])), 7638104972323651592)),
        242, ),
        ); 
}
```

### Stack overflow

The following test crashes with a stack overflow. It is possible to construct deeply nested ParamTypes which when consumed by the parser cause a stack overflow. Depending on the system configuration it might be required to change the 13500 in order to witness the crash.
The stack overflow here is caused by fuels-rs/packages/fuels-core/src/codec/abi_decoder.rs:60-63.

#### Figure 32.7: Unit test that crashes with a stack overflow.

```rust
#[test]
pub fn stack_overflow() {
    use crate::types::param_types::ParamType;
    use crate::types::param_types::ParamType::*;
    let mut typ: ParamType = U16;
    for _ in 0..13500 {
        typ = Vector(Box::new(typ));
    }
        fuzz_abi_decode(
            &[8, 9, 9, 9, 9, 9, 9, 9],
            typ
    ); 
}
```

### Memory allocation

The following unit test can panic with a failed memory allocation at fuels-rs/packages/fuels-core/src/codec/abi_decoder.rs:138. This is related to the “Capacity overflow” already mentioned in this finding. If the array length is lower than the same attack vector might lead to a hang of the parser. If it is higher, then it might lead to a capacity overflow like described above.

#### Figure 32.8: Unit test that crashed due to large memory allocation.
```rust
#[test]
pub fn capacity_maloc() {
    use crate::types::param_types::ParamType::*;
    fuzz_abi_decode(
        &[8, 8, 7, 252, 201, 8, 8, 8],
        Array(Box::new(U8), 72340198607880449)
    );
}
```
## Recommendations
Short term, avoid panicing by handling the errors appropriately. For the stack overflows, make sure to avoid recursive calls and instead implement parsing iteratively. Also, setup reasonable length limits for types which include a length like Arrays.
Long term, deploy a fuzzer for the ABI decoding parser. The above fuzz harness can be used together with Trail of Bit’s [test-fuzz](https://github.com/trailofbits/test-fuzz) fuzzer. The instructions for setting it up can be found in the documentation of the project.

### PR DESCRIPTION
**Pull request by Br1ght0ne**
Close #1108.
Thanks to @segfault-magnet for helping with debugging `stack_overflow` and the idea for `DebugWithDepth`.

### Checklist
- [x] I have linked to any relevant issues.
- [x] I have updated the documentation.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary labels.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Review by hal3e — APPROVED** [review]
LGTM. Great job!

**Review by MujkicA — DISMISSED** [review]

File: packages/fuels-core/src/codec/abi_decoder.rs [hunk 1]
```diff
@@ -510,6 +511,97 @@ mod tests {
...
+        assert!(matches!(result, Err(Error::InvalidType(_))));
+    }
+
+    #[test]
+    pub fn multiply_overflow_enum() {
+        let result = ABIDecoder::default().decode(
+            &Enum {
+                variants: EnumVariants {
+                    param_types: vec![
+                        Array(Box::new(Array(Box::new(RawSlice), 8)), 576469587185895432),
```
Comment (iqdecay) [inline packages/fuels-core/src/codec/abi_decoder.rs:526]: I think we can add a comment about what this number represents?

File: packages/fuels-core/src/codec/abi_decoder.rs [hunk 2]
```diff
@@ -510,6 +511,97 @@ mod tests {
...
+                        B256,
+                        B256,
+                        B256,
+                        B256,
+                        B256,
+                    ],
+                },
+                generics: vec![U16],
+            },
+            &[0, 8, 8, 8, 9, 8, 0, 8, 8, 8, 8, 8, 15, 8, 8, 8],
```
Comment (iqdecay) [inline packages/fuels-core/src/codec/abi_decoder.rs:541]: same here

File: packages/fuels-core/src/codec/abi_decoder/bounded_decoder.rs [hunk 3]
```diff
@@ -308,13 +311,14 @@ impl BoundedDecoder {
...
-        let enum_width = variants.compute_encoding_width_of_enum();
+        let enum_width = variants.compute_encoding_width_of_enum()?;
 
         let discriminant = peek_u32(bytes)? as u8;
         let selected_variant = variants.param_type_of_variant(discriminant)?;
 
-        let words_to_skip = enum_width - selected_variant.compute_encoding_width();
-        let enum_content_bytes = skip(bytes, words_to_skip * WORD_SIZE)?;
+        let words_to_skip = enum_width - selected_variant.compute_encoding_width()?;
+        let bytes_to_skip = words_to_skip .checked_mul(WORD_SIZE).ok_or_else(|| error!(InvalidData, "attempt to multiply words_to_skip ({words_to_skip:?}) by WORD_SIZE ({WORD_SIZE:?}) with overflow"))?;
```
Comment (iqdecay) [inline packages/fuels-core/src/codec/abi_decoder/bounded_decoder.rs:320]: I think a more user-friendly error message would be along the lines of: 
`Overflow error while decoding {type}`. The user doesn't need to be exposed to the inner details IMO.

File: packages/fuels-core/src/codec/abi_decoder/bounded_decoder.rs [hunk 4]
```diff
@@ -457,3 +461,90 @@ fn skip(slice: &[u8], num_bytes: usize) -> Result<&[u8]> {
         Ok(&slice[num_bytes..])
     }
 }
+
+/// Allows `Debug` formatting of arbitrary-depth nested `ParamTypes` by
+/// omitting the details of inner types if max depth is exceeded.
+pub(crate) struct DebugWithDepth<'a> {
```
Comment (iqdecay) [inline packages/fuels-core/src/codec/abi_decoder/bounded_decoder.rs:467]: nice

File: packages/fuels-core/src/types/param_types.rs [hunk 5]
```diff
@@ -67,7 +67,8 @@ impl ParamType {
         param_type: &ParamType,
         available_bytes: usize,
     ) -> Result<usize> {
-        let memory_size = param_type.compute_encoding_width() * WORD_SIZE;
+        let encoding_width = param_type.compute_encoding_width()?;
+        let memory_size = encoding_width.checked_mul(WORD_SIZE).ok_or_else(|| error!(InvalidData, "attempt to multiply encoding_width ({encoding_width:?}) by WORD_SIZE ({WORD_SIZE:?}) with overflow"))?;
```
Comment (iqdecay) [inline packages/fuels-core/src/types/param_types.rs:71]: same remark here regarding the error message that gives the user too much information.

File: packages/fuels-core/src/types/param_types.rs [hunk 6]
```diff
@@ -81,7 +82,10 @@ impl ParamType {
                 "{remainder} extra bytes detected while decoding heap type"
             ));
         }
-        Ok(available_bytes / memory_size)
+        let num_of_elements = available_bytes
+            .checked_div(memory_size)
+            .ok_or_else(|| error!(InvalidData, "memory_size of 0"))?;
```
Comment (iqdecay) [inline packages/fuels-core/src/types/param_types.rs:87]:
```suggestion
            .ok_or_else(|| error!(InvalidData, "Type {param_type} has a memory_size of 0"))?;
```

**Review by iqdecay — CHANGES_REQUESTED** [review]
Left some remarks regarding error messages, otherwise lgtm!

File: packages/fuels-core/src/codec/abi_decoder.rs [hunk 7]
```diff
@@ -510,6 +511,97 @@ mod tests {
...
+        );
+        assert!(matches!(result, Err(Error::InvalidData(_))));
+    }
+
+    #[test]
+    pub fn multiply_overflow_arith() {
+        let mut typ: ParamType = U16;
+        for _ in 0..50 {
+            typ = Array(Box::new(typ), 8);
+        }
```
[the remark this answers is not in the document]
Comment (Br1ght0ne) [reply packages/fuels-core/src/codec/abi_decoder.rs:571]: `typ` is being wrapped with `Array` 50 times, I think the loop always runs completely.

**Comment by Br1ght0ne** [comment]
To help with reviewing, here are the changes added in latest merge (a7a9acc):

> Extra changes:
> - `ABIDecoder.config` is now `pub`
> - `ParamType::validate_is_decodable` now uses `DebugWithDepth` (with
depth 0 in tests and `decoder.max_depth` in use)
> - `DebugWithDepth` moved to `utils` module

File: packages/fuels-core/src/types/param_types.rs [hunk 8]
```diff
@@ -1466,18 +1559,19 @@ mod tests {
             ParamType::Bool,
         ];
         let param_types_no_vector = vec![ParamType::U64, ParamType::U32];
-        let nested_heap_type_error_message = "Invalid type: Nested heap types are currently not \
+        let nested_heap_type_error_message =
+            "Invalid type: type ... is not decodable: nested heap types are currently not \
```
Comment (iqdecay) [inline packages/fuels-core/src/types/param_types.rs:1563]: same here.

File: packages/fuels-programs/src/call_utils.rs [hunk 9]
```diff
@@ -358,42 +352,22 @@ pub(crate) fn get_single_call_instructions(
...
-            // We know a Vec/Bytes/String struct has its third WORD contain the length of the
-            // underlying vector, so use a 2 offset to store the length in 0x16, which is a free
-            // register.
-            op::lw(0x16, RegId::RET, 2),
-            // The in-memory size of the type is (in-memory size of the inner type) * length
-            op::muli(0x16, 0x16, inner_type_byte_size as u16),
-            op::retd(0x15, 0x16),
-        ]);
-    }
+    instructions.extend(extract_heap_data(output_param_type)?);
```
Comment (iqdecay) [inline packages/fuels-programs/src/call_utils.rs:355]: I think those comments are essential to understand the bytecode mumbo-jumbo needed there.

**Review by iqdecay — CHANGES_REQUESTED** [review]
You removed comments that are useful IMO. Also some imprecisions in the error message.

File: packages/fuels-programs/src/call_utils.rs (hunk 9 above)
Comment (Br1ght0ne) [reply packages/fuels-programs/src/call_utils.rs:355]: Code was changed in a merge, this no longer stands as those comments were moved somewhere else.

File: packages/fuels-core/src/types/param_types.rs [hunk 10]
```diff
@@ -81,7 +84,10 @@ impl ParamType {
                 "{remainder} extra bytes detected while decoding heap type"
             ));
         }
-        Ok(available_bytes / memory_size)
```
Comment (digorithm) [inline packages/fuels-core/src/types/param_types.rs:84]: Damn, that's an excellent catch. Super silly in hindsight, though.

**Review by digorithm — APPROVED** [review]
Excellent work, @Br1ght0ne!

**Review by hal3e — APPROVED** [review]
:rocket:

**Pull request merged**